File: dns_server/dns_server.cpp
#include "./modules/dns_module.h"
#include "./modules/connectsock.h"
#include <stdexcept>
#include <string>
#include <sys/socket.h>
#include <sys/types.h>
#include <fstream>

const char* RECORD_FILE = "./modules/dns_records.txt"; // records file for local lookups
std::vector<DnsRecord> RECORDS; // where local records will be stored
IPv4Addr AROOT = IPv4Addr("198.41.0.4"); // this is a.root-servers.net

// Parses a file with records and stores them in a vector
// record example: tunnel.mydomain.top IN NS ns.mydomain.top
std::vector<DnsRecord> records_from_file(const char* file)
{
    std::vector<DnsRecord> result; 

    // opening the file
    std::ifstream fin(file);
    char reader[500]; // a buffer to read into
    if (!fin.is_open())
    {
        throw std::runtime_error("Could not open records file!");
    }

    while (fin >> reader) { // domain
        DnsRecord record;
        record.domain = reader;
        fin >> reader; // class
        fin >> reader; // qtype
        record.qtype = QueryTypeFunc::from_string(reader);

        fin.getline(reader, sizeof(reader)-1); // data
        get_byte_array_from_string(record.value, reader);

        // trim white spces
        record.value.erase(record.value.begin(), std::find_if(record.value.begin(), record.value.end(), [](unsigned char ch) {
            return !std::isspace(ch);
        }));

        record.ttl = 300;
        result.push_back(record);
    }

    fin.close();
    return result;
}

// Searches for a domain with a specific type within a vector of records
std::vector<DnsRecord> search_in_records(std::string qname, QueryType qtype, std::vector<DnsRecord> &records)
{
    std::vector<DnsRecord> result;

    for (auto &rec : records)
    {
        if (strcmp(rec.domain.c_str(), qname.c_str()) == 0) {
            // the domain is the one we are looking for
            if (rec.qtype == QueryType::NS || rec.qtype == qtype) // checking for the same type or maybe an NS
                result.push_back(rec);
        }
        else if (qname.ends_with(rec.domain)) // if the domain is ending whith the string we aim for an NS 
            if (rec.qtype == QueryType::NS)
                result.push_back(rec);
    }

    return result;
}

// Performs a lookup on the RECORDS vector for a domain and a type
DnsPacket local_lookup(std::string qname, QueryType qtype)
{
    DnsPacket result;
    std::vector<DnsRecord> records = search_in_records(qname, qtype, RECORDS);

    if (records.empty())
        result.header.rescode = ResultCode::NXDOMAIN; // nothing found
    else {
        for (auto &rec : records) 
        {
            if (qname == rec.domain && qtype == rec.qtype) // if everything is matching we have an answer
            {
                result.answers.push_back(rec);
                result.header.answers ++;
            }
            else // otherwise there is an NS that we can ask
            {
                result.header.authoritative_entries ++;
                result.authorities.push_back(rec); 
            }
        }
    }

    return result;
}

// Queryes a DNS server for a specific domain and type
DnsPacket lookup(std::string qname, QueryType qtype, IPv4Addr ns)
{
    // connection stuff
    struct sockaddr_in server_addr;
    bzero(&server_addr, sizeof(server_addr)); // clearing
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(53);
    inet_pton(AF_INET, ns.to_string().c_str(), &server_addr.sin_addr);
    
    // this socket is for listening for responses
    int socket = passive_sock_udp("43210"); // picking a random port to send the request
    if (set_socket_timeout(socket, 5) < 0)
    {
        std::cout << "Warning: Error when setting socket timeout" << std::endl;
    }

    DnsPacket packet = DnsPacket(); // the packet that will be sent

    try{
        // crafting the header
        packet.header.id = 1234; // random id
        packet.header.questions = 1;
        packet.header.recursion_desired = true;

        // crafting the body
        packet.questions.push_back(DnsQuestion(qname, qtype));

        // converting the packet into a buffer
        BytePacketBuffer request_buffer = BytePacketBuffer(); // request buffer
        packet.write(request_buffer);

        const size_t request_size = request_buffer.getPos(); // pakcet size

        // char request_bytes[p_size];
        // std::copy(request_buffer.buf.begin(), request_buffer.buf.end(), request_bytes);

        // sending the request
        ssize_t bytes_sent = sendto(socket, request_buffer.buf.data(), request_size, 0, (struct sockaddr*) &server_addr, sizeof(server_addr));
        if (bytes_sent <= 0)
        {
            close(socket);
            throw std::runtime_error("Some error occurred when sending data!\n");
        }

        // receiving response
        char response_bytes[BytePacketBuffer::packet_size]; // response goes here
        ssize_t bytes_read = recv(socket, response_bytes, BytePacketBuffer::packet_size, 0); 
        if (bytes_read <= 0)
        {
            close(socket);
            throw std::runtime_error("Some error occurred when receiving data!\n");
        }
        
        BytePacketBuffer response_buffer = BytePacketBuffer(response_bytes, bytes_read);
        
        close(socket);
        return DnsPacket::from_buffer(response_buffer);

    }
    catch (std::runtime_error e)
    {
        close(socket);
        throw e;
    }
}

// Tries to resolve a domain name through recursive lookups
DnsPacket recursive_lookup(std::string qname, QueryType qtype)
{
    IPv4Addr ns = IPv4Addr("127.0.0.1"); // first we try a local lookup
    DnsPacket response;

    while (true)
    {

        try {
            
            if (ns.to_string() == "127.0.0.1") {
                // first perform a local lookup to see if there is a local domain assigned
                std::cout << "Attepmpting LOCAL lookup of " << QueryTypeFunc::to_string(qtype) << " " << qname << std::endl;
                response = local_lookup(qname, qtype);
            }
            else {
                std::cout << "Attepmpting ONLINE lookup of " << QueryTypeFunc::to_string(qtype) << " " << qname << " with NS " << ns.to_string() << std::endl;
                response = lookup(qname, qtype, ns);
            }
            
        }
        catch (std::runtime_error e)
        {
            throw e;
        }
        
        if (response.header.rescode == ResultCode::NXDOMAIN && ns.to_string() == "127.0.0.1") // no domain found locally
        {
            // local lookup failed, trying to ask a-root
            ns = AROOT;
            continue;
        }

        // success
        if (!response.answers.empty() && response.header.rescode == ResultCode::NOERROR)
            return response;

        // domain not found
        if (response.header.rescode == ResultCode::NXDOMAIN)
            return response;
        
        // if we are here it means that there are some resources we can check
        IPv4Addr new_ns = response.get_resolved_ns(qname);
        if (new_ns.to_string() != "0.0.0.0"){ // if the ip is 0.0.0.0 it means there are no entries
            ns = new_ns;
            continue;
        }
        
        std::vector<std::string> new_ns_names = response.get_unresolved_ns(qname);
        if (new_ns_names.empty())
            return response;

        // trying to get the ip of the unresolved ns
        DnsPacket recursive_response = recursive_lookup(new_ns_names.front(), QueryType::A);

        new_ns = recursive_response.get_random_a();

        if (new_ns.to_string() != "0.0.0.0") // if the ip is 0.0.0.0 it means there are no entries
            ns = new_ns;
        else
            return response;
    }
}

// The main function that handles a request
void handle_query(int socket)
{
    struct sockaddr_storage source_addr;
    socklen_t source_addrlen = sizeof(source_addr);
    char request_bytes[BytePacketBuffer::packet_size];

    // Waiting for data
    ssize_t bytes_read = recvfrom(socket, request_bytes, BytePacketBuffer::packet_size, 0, (struct sockaddr*) &source_addr, &source_addrlen);
    if (bytes_read <= 0)
    {
        throw std::runtime_error("An error occured when receiving data! (handle_query)");
    }

    try
    {
        // Processing the data
        BytePacketBuffer request_buffer = BytePacketBuffer(request_bytes, (size_t) bytes_read);
        DnsPacket request = DnsPacket::from_buffer(request_buffer);

        DnsPacket packet; // response packet
        packet.header.id = request.header.id;
        packet.header.recursion_desired = true;
        packet.header.recursion_available = true;
        packet.header.response = true;

        if (request.header.questions == 1 && request.questions.size() == 1)
        {
            DnsQuestion question = request.questions.front();
            std::cout << "Received query:\n" << question.to_string() << std::endl;

            DnsPacket result;
            try{
                // perform a recursive lookup
                result = recursive_lookup(question.name, question.qtype);
                packet.header.rescode = result.header.rescode;

            }
            catch(std::runtime_error e)
            {
                std::cout << e.what() << std::endl;
                packet.header.rescode = ResultCode::SERVFAIL;
            }

            // building the packet's body
            packet.questions.push_back(question);

            std::cout << "Response packet:" << std::endl;
            for (size_t i = 0; i < result.answers.size(); i ++)
            {
                if (i == 0)
                    std::cout << "Answer:\n";
                std::cout << result.answers[i].to_string() << std::endl;
                packet.answers.push_back(result.answers[i]);
            }
            for (size_t i = 0; i < result.authorities.size(); i ++)
            {
                if (i == 0)
                    std::cout << "Authorities:\n";
                std::cout << result.authorities[i].to_string() << std::endl;
                packet.authorities.push_back(result.authorities[i]);
            }
            for (size_t i = 0; i < result.resources.size(); i ++)
            {
                if (i == 0)
                    std::cout << "Resources:\n";
                std::cout << result.resources[i].to_string() << std::endl;
                packet.resources.push_back(result.resources[i]);
            }
        }
        else{
            packet.header.rescode = ResultCode::FORMERR;
        }

        // Sending data back
        BytePacketBuffer response_buffer;

        packet.write(response_buffer);

        const size_t response_size = response_buffer.getPos();
        // char response_bytes[datalen];
        // std::copy(response_buffer.buf.begin(), response_buffer.buf.end(), response_bytes);

        ssize_t bytes_sent = sendto(socket, response_buffer.buf.data(), response_size, 0, (struct sockaddr*) &source_addr, source_addrlen);

        if (bytes_sent <= 0)
        {
            throw std::runtime_error("An error occured when sending data! (handle_query)");
        }
    }
    catch (std::runtime_error e)
    {
        std::cout << strerror(errno) << std::endl;
        throw e;
    }
}

int main(int argc, char** argv)
{
    if (argc != 2)
    {
        std::cout << "Usage: " << argv[0] << " port" << std::endl;
        return 0;
    }

    // getting the local records
    RECORDS = records_from_file(RECORD_FILE);

    // Start listening
    int socket = passive_sock_udp(argv[1]);
    if (socket < 0){
        return -1;
    }

    std::cout << "Listening on 0.0.0.0:" << argv[1] << " ..." << std::endl;

    while (true)
    {
        try {
            handle_query(socket);
        }
        catch (std::runtime_error e) {
            std::cout << e.what() << std::endl;
        }
    }

    return 0;
}

File: dns_server/dns_server.py
import socket
from scapy.all import *
from scapy.layers.dns import DNS, DNSQR, DNSRR
from scapy.layers.inet import IP, UDP
import sys
import random
from loguru import logger as logging
from modules import dns_module

PACKET_SIZE = 1024
AROOT = "198.41.0.4" # a.root-servers.net
RECORDS = list()
RECORDS_FILE = "./modules/dns_records.txt"

# loguru configurations
logging.remove()
logging.add(sys.stdout, colorize=True, format="<level>{level.icon}</level> <level>{message}</level>", level="TRACE")
logging.level(name='INFO', icon='[~]', color='<cyan><bold>')
logging.level(name='WARNING', icon='[!]', color='<yellow><bold>')
logging.level(name='DEBUG', icon='[*]', color='<blue><bold>')
logging.level(name='ERROR', icon='[X]', color='<red><bold>')
logging.level(name='SUCCESS', icon='[#]', color='<green><bold>')

# Parses a file with records and stores them in a list
# record example: tunnel.mydomain.top IN NS ns.mydomain.top
def records_from_file(file):
    result = list()

    with open(file, "r") as f:
        for line in f.readlines():
            tokens = line.split() # dns record: ns.mydomain.top IN A 10.8.8.16
                                  # tokens: [domain, class, type, data]
            result.append(DNSRR(rrname=tokens[0], rclass=1, type=dns_module.type_number_from_string(tokens[2]), rdata=tokens[3], ttl=300))

    return result

# Searches for a domain with a specific type within a vector of records
def search_in_records(qname, qtype, records):
    result = list()

    for record in records:
        domain = record.rrname.decode()
        if qname == domain: # looking for the same domain
            # looking for the same type or for NS type
            if qtype == record.type or record.type == 2: # NS has type == 2
                result.append(record)
        elif (qname[-len(domain):] == domain): # if the domain is ending whith the string we aim for an NS 
            if (record.type == 2):
                result.append(record)

    return result


# Performs a lookup on the RECORDS vector for a domain and a type
def local_lookup(qname, qtype):
    result = DNS(ancount=0, nscount=0, arcount=0)
    result.rcode = dns_module.response_code_to_number("NOERROR")  # Default to NOERROR
    records = search_in_records(qname, qtype, RECORDS)

    # Process records
    for record in records:
        rrname = record.rrname.decode()
        rtype = record.type

        if qname == rrname and qtype == rtype:
            # Record matches query
            if result.ancount == 0:
                result.an = record
            else:
                result.an = result.an / record
            result.ancount += 1
        else:
            # Process NS or other records
            if rtype == 2:  # Assuming type 2 is NS
                if result.nscount == 0:
                    result.ns = record
                else:
                    result.ns = result.ns / record
                result.nscount += 1
            elif rtype == 1:  # Assuming type 1 is A record
                if result.arcount == 0:
                    result.ar = record
                else:
                    result.ar = result.ar / record
                result.arcount += 1

    # If no answers, set NXDOMAIN
    if result.ancount == 0 and result.nscount == 0:
        result.rcode = dns_module.response_code_to_number("NXDOMAIN")

    return result

# Queryes a DNS server for a specific domain and type
def lookup(qname, qtype, ns):
    # this socket is for listening for responses
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.settimeout(3)
    sock.bind(('', 43210)) # picking a random port to send the request

    packet = DNS(id=1234, rd=1, qd=DNSQR(qname=qname, qtype=qtype))
    request_bytes = bytes(packet)

    try:
        # sending the request
        sock.sendto(request_bytes, (ns, 53))
        # receiving a response
        response_bytes, _ = sock.recvfrom(PACKET_SIZE)
        response = DNS(response_bytes)
        sock.close()
        return response
    except Exception as e:
        sock.close()
        raise RuntimeError(f"Some error occurred when sending/receiving data! {e}") from e

# Tries to resolve a domain name through recursive lookups
def recursive_lookup(qname, qtype):
    ns = "127.0.0.1" # first we try a local lookup
    response = None

    try:
        while True:
            if ns == "127.0.0.1":
                # first perform a local lookup to see if there is a local domain assigned
                logging.info(f"Attempting LOCAL lookup of {dns_module.type_number_to_string(qtype)} {qname}")
                response = local_lookup(qname, qtype)
            else:
                logging.info(f"Attempting ONLINE lookup of {dns_module.type_number_to_string(qtype)} {qname} with NS {ns}")
                response = lookup(qname, qtype, ns)

            if response.rcode == 3 and ns == "127.0.0.1": # no domain found locally
                ns = AROOT
                continue
            
            # success
            if response and response.rcode == 0 and response.ancount > 0:
                return response

            # domain not found
            if response.rcode == 3:  # NXDOMAIN
                return response
            
            # if we are here it means that there are some resources we can check
            new_ns = None
            new_ns = dns_module.get_resolved_ns(response, qname)

            if new_ns:
                ns = new_ns 
                continue 
            
            new_ns_names = dns_module.get_unresolved_ns(response, qname)
            if (len(new_ns_names) == 0):
                break

            recursive_response = recursive_lookup(new_ns_names[random.randint(0, len(new_ns_names) - 1)], 1) # trying to resolve the first unresolved ns

            if recursive_response.ancount == 0:
                break 
            else:
                ns = recursive_response.an.rdata
            
    except RuntimeError as e:
        logging.error(f"recursive_lookup error: {e}")
        raise e 
    except AttributeError as e:
        logging.error(f"recursive_lookup error: {e}")
        raise e 
    except Exception as e:
        logging.error(f"recursive_lookup error: {e}")
        raise e 

    return response


# The main function that handles a request
def handle_query(sock):
    data, addr = sock.recvfrom(PACKET_SIZE)
    request = DNS(data)

    if request.opcode == 0 and request.qr == 0:  # Standard query
        qname = request.qd.qname.decode()
        qtype = request.qd.qtype

        logging.info(f"Received query for {qname} type {dns_module.type_number_to_string(qtype)}")

        response = DNS(id=request.id, qr=1, rd=request.rd, ra=1, qd=request.qd, qdcount=request.qdcount)

        try:
            result = recursive_lookup(qname, qtype)
            response.rcode = result.rcode
            response.ancount = result.ancount
            response.nscount = result.nscount
            response.arcount = result.arcount
            response.an = result.an
            response.ns = result.ns
            response.ar = result.ar
        except RuntimeError as e:
            response.rcode = 2  # SERVFAIL
        except AttributeError as e:
            response.rcode = 2  # SERVFAIL
        except Exception as e:
            response.rcode = 2  # SERVFAIL
    
    sock.sendto(bytes(response), addr)

    logging.info("Query handeled!")
    

if __name__ == "__main__":
    if len(sys.argv) != 2:
        logging.debug(f"Usage: {sys.argv[0]} port")
        sys.exit(0)

    RECORDS = records_from_file(RECORDS_FILE)

    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        # start listening
        sock.bind(('0.0.0.0', int(sys.argv[1])))
    except Exception as e:
        logging.error(f"main error: {e}")
        sys.exit(0)

    logging.info(f"DNS server listening on port {sys.argv[1]}")

    while True:
        try:
            handle_query(sock)
        except KeyboardInterrupt:
            logging.warning("Shutting down DNS server")
            sock.close()
            exit()
        except RuntimeError as e:
            logging.error(f"main error: {e}")
            logging.warning("Shutting down DNS server")
            sock.close()
            exit()
